### ISSUE TITLE
Batch Inference to get bounding box coordinates using TensorRT

### ISSUE THREAD

**Issue by santoshmedisetty** [ultralytics/yolov5#6260]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi, I checked the issues and discussions and couldn't find a proper resource.  I converted my custom yolov5n .pt model to .engine model using export.py. Now I want to perform batch inference on images to get the bounding box coordinates. Can anyone please help me or guide me to a resource?

**Comment by glenn-jocher** [comment]
See export.py argparser arguments for exporting at arbitrary batch sizes, i.e. --batch 64.

```python
!pip install -U nvidia-tensorrt --index-url https://pypi.ngc.nvidia.com  # install
!python export.py --weights yolov5s.pt --include engine --imgsz 640 640 --device 0  # export
!python detect.py --weights yolov5s.engine --imgsz 640 640 --device 0  # inference
```

**Comment by santoshmedisetty** [comment]
Thanks @glenn-jocher for the prompt reply. I exported the model to .engine by specifying the required batch size. I was looking for batch inference solution similar to https://github.com/ultralytics/yolov5/issues/1806 to get the bounding box coordinates.

**Comment by glenn-jocher** [comment]
@santoshmedisetty you can use your .engine model in detect.py, but this method usually uses batch-size 1 dataloaders. If you load your engine model with PyTorch Hub you can use any batch size:

YOLOv5 🚀 [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) models allow for simple model loading and inference in a pure python environment without using `detect.py`. 

### Simple Inference Example

This example loads a pretrained YOLOv5s model from PyTorch Hub as `model` and passes an image for inference. `'yolov5s'` is the lightest and fastest YOLOv5 model. For details on all available models please see the [README](https://github.com/ultralytics/yolov5#pretrained-checkpoints).
```python
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')  # or yolov5m, yolov5l, yolov5x, custom

# Images
im = 'https://ultralytics.com/images/zidane.jpg'  # or file, Path, URL, PIL, OpenCV, numpy, list

# Inference
results = model(im)

# Results
results.print  # or .show(), .save(), .crop(), .pandas(), etc.

results.xyxy[0]  # im predictions (tensor)
results.pandas().xyxy[0]  # im predictions (pandas)
#      xmin    ymin    xmax   ymax  confidence  class    name
# 0  749.50   43.50  1148.0  704.5    0.874023      0  person
# 2  114.75  195.75  1095.0  708.0    0.624512      0  person
# 3  986.00  304.00  1028.0  420.0    0.286865     27     tie
```

See [YOLOv5 PyTorch Hub Tutorial](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) for details.

Good luck 🍀 and let us know if you have any other questions!

**Comment by santoshmedisetty** [comment]
Hi @glenn-jocher, First of all, a big thank you for helping me. I loaded my .engine model using  pytorch hub but got an error while making predictions(result = model(inputs)). Both my inputs and the model are on the same device(GPU). Please see the below error message. 

RuntimeError: CUDA error: an illegal memory access was encountered
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
[01/10/2022-19:04:40] [TRT] [E] 1: [defaultAllocator.cpp::deallocate::35] Error Code 1: Cuda Runtime (an illegal memory access was encountered)
[01/10/2022-19:04:40] [TRT] [E] 1: [defaultAllocator.cpp::deallocate::35] Error Code 1: Cuda Runtime (an illegal memory access was encountered)

P.S: I used 'torch.cuda.current_device()' at the beginning of my code

**Comment by glenn-jocher** [comment]
@santoshmedisetty I'll take a look. I haven't actually tried to run TRT via torch hub models yet.

**Comment by glenn-jocher** [comment]
@santoshmedisetty I'm able to reproduce an error in Colab using this code (after running Setup cell):

```python
!pip install -U nvidia-tensorrt --index-url https://pypi.ngc.nvidia.com  # install
!python export.py --weights yolov5s.pt --include engine --imgsz 640 640 --device 0  # export
!python detect.py --weights yolov5s.engine --imgsz 640 640 --device 0  # inference

# PyTorch Hub
import torch
model = torch.hub.load('ultralytics/yolov5', 'custom', 'yolov5s.engine')
model('https://ultralytics.com/images/zidane.jpg').print()
```

I'll add a TODO here to look into this but I can't provide a timeline for a fix currently. I've verified detect.py inference works correctly (per reproduction code above).

TODO: TRT PyTorch Hub inference fix

**Comment by glenn-jocher** [comment]
TODO: TRT PyTorch Hub inference fix

**Comment by santoshmedisetty** [comment]
Thank you @glenn-jocher. I'll wait for the fix.

**Comment by glenn-jocher** [comment]
@santoshmedisetty the issue seems to originate with a missing LD_LIBRARY_PATH variable, i.e. depending on your installation:
```python
%env LD_LIBRARY_PATH=/usr/local/cuda-11.1/lib64:/content/cuda-11.1/lib64:/content/TensorRT-8.2.0.6/lib:/usr/local/nvidia/lib:/usr/local/nvidia/lib64  # add to path
```

See https://forums.developer.nvidia.com/t/importerror-libnvinfer-so-4-cannot-open-shared-object-file-no-such-file-or-directory/62243/4

**Comment by glenn-jocher** [comment]
@santoshmedisetty the error is reproducible very simple with:
```python
import tensorrt
```

**Comment by glenn-jocher** [comment]
@santoshmedisetty ok so it seems the above info only holds for the tar install. For the new pip install instructions there is a different error only on hub inference (detect.py is fine):
```python
# TensorRT 
# https://docs.nvidia.com/deeplearning/tensorrt/install-guide/index.html#installing-pip
!pip install -U nvidia-tensorrt --index-url https://pypi.ngc.nvidia.com  # install
!python export.py --weights yolov5s.pt --include engine --imgsz 640 640 --device 0  # export
!python detect.py --weights yolov5s.engine --imgsz 640 640 --device 0  # inference

# PyTorch Hub
import torch
model = torch.hub.load('ultralytics/yolov5', 'custom', 'yolov5s.engine')
model('https://ultralytics.com/images/zidane.jpg').print()
```

**Comment by santoshmedisetty** [comment]
@glenn-jocher  So, do you mean, I'm getting the error because of TensorRT installation? I'm running my code on Ubuntu 20.10. And installed tensorrt using pip

**Comment by glenn-jocher** [comment]
@santoshmedisetty error with pip install is 

```bash
RuntimeError: CUDA error: an illegal memory access was encountered
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
```

<img width="1067" alt="Screen Shot 2022-01-10 at 5 00 33 PM" src="https://user-images.githubusercontent.com/26833433/148873832-a2c223c0-6da5-435b-afa6-c829d77c99b9.png">

**Comment by santoshmedisetty** [comment]
Yes @glenn-jocher. I got the same error.

**Comment by santoshmedisetty** [comment]
Is this something that can be fixed by changing the installation? Or, any change in export.py can fix this?

**Comment by glenn-jocher** [comment]
@santoshmedisetty export and inference work correctly with export.py and detect.py, so the error seems isolated to torch hub. I'm not sure what the cause is. If you discover anything helpful please let us know!

**Comment by santoshmedisetty** [comment]
@glenn-jocher Sure. Will try. Meanwhile, if you find anything that solves the problem, please let me know

**Comment by zhiqwang** [comment]
> So, do you mean, I'm getting the error because of TensorRT installation? I'm running my code on Ubuntu 20.10. And installed tensorrt using pip

Hi @santoshmedisetty , I guess you can try the following:
```
pip install -U nvidia-tensorrt --index-url https://pypi.ngc.nvidia.com
```
Check the docs https://docs.nvidia.com/deeplearning/tensorrt/install-guide/index.html#installing-pip and https://docs.ultralytics.com/yolov5/tutorials/model_export#issuecomment-1008604368 for more details.

**Comment by santoshmedisetty** [comment]
Hi @zhiqwang, The issue is not with the tensorrt installation. For batch inference, when I give a list of images as input to the model for example,

I'm using the .engine model here for inference.

frames = [img1, img2, img3,...]

results = model(frames)

<img width="1279" alt="Screen Shot 2022-01-17 at 11 22 15 AM" src="https://user-images.githubusercontent.com/28855704/149796776-eb303345-dd4c-48d7-9934-8979166ea574.png">

Here img1, img2,... are images in the form of numpy arrays.

Instead if I give the input to the model as tensor of tensors, I get the below error.

<img width="1279" alt="Screen Shot 2022-01-17 at 11 27 13 AM" src="https://user-images.githubusercontent.com/28855704/149797397-4c3072b1-6b77-4fff-9590-afe5bd20b193.png">


For the .pt model, I used the inputs as a list of images as np arrays and it worked perfectly fine.

**Comment by zhiqwang** [comment]
@santoshmedisetty , 

> The issue is not with the tensorrt installation.

Got it! I will try to reproduce your question here this week.

And just FYI, I guess you can specify the CUDA device such with `CUDA_VISIBLE_DEVICES=0 python detect.py ...`.

**Comment by santoshmedisetty** [comment]
@zhiqwang Thank you. I already specified the device using torch.cuda.current_device()

**Comment by UygarUsta99** [comment]
Any updates to this ? I need to batch inference some images using tensorrt and having some trouble atm.

**Comment by glenn-jocher** [comment]
@UygarUsta99 @santoshmedisetty @zhiqwang good news 😃! Your original issue may now be fixed ✅ in PR #7560 implementing a TRT PyTorch Hub inference fix by @youngjae-avikus. 

<img width="941" alt="Screen Shot 2022-04-24 at 12 44 49 PM" src="https://user-images.githubusercontent.com/26833433/164993866-ae017d88-5f09-41be-bc2c-2186cd43986f.png">

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!
